File: BionicLimbSimulation_unity/Assets/Tests/PlayMode/PlaybackIntegrationTest.cs
using System.Collections;
using System.IO;
using NUnit.Framework;
using UnityEngine;
using UnityEngine.TestTools;
using BionicLimb.Core;
using BionicLimb.Hand;
using BionicLimb.Classification;

namespace BionicLimb.Tests
{
    /// <summary>
    /// PlayMode integration test: verifies the full pipeline from file load
    /// through classification and pose output over real frames.
    /// </summary>
    public class PlaybackIntegrationTest
    {
        private GameObject _controllerGO;
        private PlaybackController _controller;
        private string _tempCsvPath;

        [UnitySetUp]
        public IEnumerator SetUp()
        {
            // Build minimal scene
            _controllerGO = new GameObject("PlaybackController");
            _controller = _controllerGO.AddComponent<PlaybackController>();

            // Create a GesturePoser with stub HandRig
            var handGO = new GameObject("Hand_Root");
            var handRig = handGO.AddComponent<HandRig>();
            handRig.fingerBones = new Transform[14];
            for (int i = 0; i < 14; i++)
            {
                var bone = new GameObject($"Bone_{i}").transform;
                bone.SetParent(handGO.transform);
                handRig.fingerBones[i] = bone;
            }

            var library = ScriptableObject.CreateInstance<GesturePoseLibrary>();
            library.poses = new GesturePose[8];
            for (int i = 0; i < 8; i++)
            {
                var pose = ScriptableObject.CreateInstance<GesturePose>();
                pose.targetRotations = new Quaternion[GesturePose.BoneCount];
                for (int b = 0; b < GesturePose.BoneCount; b++)
                    pose.targetRotations[b] = Quaternion.identity;
                library.poses[i] = pose;
            }

            var poser = handGO.AddComponent<GesturePoser>();
            poser.handRig = handRig;
            poser.poseLibrary = library;

            _controller.gesturePoser = poser;

            // Write a minimal .myo.csv to a temp file
            _tempCsvPath = Path.Combine(Application.temporaryCachePath, "integration_test.myo.csv");
            WriteSyntheticCsv(_tempCsvPath, samples: 200, sampleRate: 200);

            yield return null;
        }

        [UnityTearDown]
        public IEnumerator TearDown()
        {
            Object.DestroyImmediate(_controllerGO);
            if (File.Exists(_tempCsvPath)) File.Delete(_tempCsvPath);
            yield return null;
        }

        // ── Tests ─────────────────────────────────────────────────────────────

        [UnityTest]
        public IEnumerator LoadFile_TransitionsToLoadedState()
        {
            _controller.LoadFile(_tempCsvPath);
            yield return null;

            Assert.AreEqual(PlaybackController.PlaybackState.Loaded, _controller.State,
                $"Expected Loaded, got {_controller.State}: {_controller.ErrorMessage}");
        }

        [UnityTest]
        public IEnumerator Play_AdvancesSampleIndex()
        {
            _controller.LoadFile(_tempCsvPath);
            yield return null;

            _controller.Play();
            yield return new WaitForSeconds(0.1f); // allow several samples to process

            Assert.Greater(_controller.CurrentSampleIndex, 0, "Sample index should have advanced.");
        }

        [UnityTest]
        public IEnumerator Play_Pause_StopsAdvancing()
        {
            _controller.LoadFile(_tempCsvPath);
            yield return null;

            _controller.Play();
            yield return new WaitForSeconds(0.05f);

            _controller.Pause();
            int indexAtPause = _controller.CurrentSampleIndex;
            yield return new WaitForSeconds(0.1f);

            Assert.AreEqual(indexAtPause, _controller.CurrentSampleIndex, "Index should freeze after Pause.");
        }

        [UnityTest]
        public IEnumerator ClassifierMode_IsSet_AfterLoad()
        {
            _controller.LoadFile(_tempCsvPath);
            yield return null;

            Assert.IsNotEmpty(_controller.ClassifierMode, "ClassifierMode should be set after load.");
        }

        [UnityTest]
        public IEnumerator CurrentResult_IsSet_AfterPlayback()
        {
            _controller.LoadFile(_tempCsvPath);
            yield return null;
            _controller.Play();
            yield return new WaitForSeconds(0.15f);

            Assert.IsNotNull(_controller.CurrentResult, "CurrentResult should be populated.");
            Assert.IsNotNull(_controller.CurrentResult.GestureName);
        }

        // ── CSV helper ────────────────────────────────────────────────────────

        private static void WriteSyntheticCsv(string path, int samples, int sampleRate)
        {
            using var w = new StreamWriter(path);
            w.WriteLine("#FORMAT_VERSION,1");
            w.WriteLine($"#SAMPLE_RATE_HZ,{sampleRate}");
            w.WriteLine("#NUM_CHANNELS,8");
            w.WriteLine("#SUBJECT_ID,integration_test");
            w.WriteLine("#GESTURE_LABELS,rest,fist,open_hand,pinch,point,thumbs_up,peace,ok");
            w.WriteLine($"#DURATION_S,{(double)samples / sampleRate:F3}");
            w.WriteLine("timestamp_ms,ch0,ch1,ch2,ch3,ch4,ch5,ch6,ch7,gesture_id,gesture_label");
            for (int i = 0; i < samples; i++)
            {
                int gid = (i * 8) / samples; // sweep through all 8 gestures
                string glabel = new[] { "rest","fist","open_hand","pinch","point","thumbs_up","peace","ok" }[gid];
                float v = gid > 0 ? 0.6f : 0.05f;
                w.WriteLine($"{i * (1000 / sampleRate)},{v},{v},{v},{v},{v},{v},{v},{v},{gid},{glabel}");
            }
        }
    }
}

File: bionic-limb-threejs/tests/PlaybackIntegration.test.ts
// @vitest-environment jsdom
import { describe, it, expect, beforeEach } from 'vitest';
import { PlaybackController } from '../src/playback/PlaybackController.js';

const VALID_CSV = `# FORMAT_VERSION,1.0
# SAMPLE_RATE_HZ,200
# NUM_CHANNELS,8
# SUBJECT_ID,TEST
# GESTURE_LABELS,rest
# DURATION_S,1.0
timestamp_ms,ch0,ch1,ch2,ch3,ch4,ch5,ch6,ch7,gesture_label,gesture_id
0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,rest,0
5.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,rest,0
10.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,rest,0
`;

describe('PlaybackController integration', () => {
  let controller: PlaybackController;

  beforeEach(() => {
    controller = new PlaybackController();
  });

  it('loadFile -> state Loaded', () => {
    controller.loadFile(VALID_CSV);
    expect(controller.state).toBe('Loaded');
  });

  it('play + update advances sample index', () => {
    controller.loadFile(VALID_CSV);
    controller.play();
    controller.update(1 / 200);
    expect(controller.currentSampleIndex).toBeGreaterThan(0);
  });

  it('pause stops advancement', () => {
    controller.loadFile(VALID_CSV);
    controller.play();
    controller.pause();
    const idx = controller.currentSampleIndex;
    controller.update(1 / 200);
    expect(controller.currentSampleIndex).toBe(idx);
  });

  it('default classifierMode is RuleBased', () => {
    expect(controller.classifierMode).toBe('RuleBased');
  });

  it('play + update sets currentResult', () => {
    controller.loadFile(VALID_CSV);
    controller.play();
    controller.update(1 / 200);
    expect(controller.currentResult).not.toBeNull();
  });

  it('gestureChanged event has detail.result with GestureResult shape', () => {
    controller.loadFile(VALID_CSV);
    controller.play();
    let capturedResult: unknown = null;
    controller.addEventListener('gestureChanged', (e: Event) => {
      capturedResult = (e as CustomEvent).detail.result;
    });
    controller.update(1 / 200);
    expect(capturedResult).not.toBeNull();
    const result = capturedResult as Record<string, unknown>;
    expect(typeof result.gestureId).toBe('number');
    expect(typeof result.gestureName).toBe('string');
    expect(typeof result.confidence).toBe('number');
    expect((result.confidence as number)).toBeGreaterThanOrEqual(0);
    expect((result.confidence as number)).toBeLessThanOrEqual(1);
    expect(result.allProbabilities).toBeInstanceOf(Float32Array);
  });

  it('progressChanged event has detail.progress in [0, 1]', () => {
    controller.loadFile(VALID_CSV);
    controller.play();
    let capturedProgress: number | null = null;
    controller.addEventListener('progressChanged', (e: Event) => {
      capturedProgress = (e as CustomEvent<{ progress: number }>).detail.progress;
    });
    controller.update(1 / 200);
    expect(capturedProgress).not.toBeNull();
    expect(capturedProgress as unknown as number).toBeGreaterThanOrEqual(0);
    expect(capturedProgress as unknown as number).toBeLessThanOrEqual(1);
  });

  it('stateChanged event has detail.state as PlaybackState string', () => {
    const states: string[] = [];
    controller.addEventListener('stateChanged', (e: Event) => {
      states.push((e as CustomEvent<{ state: string }>).detail.state);
    });
    controller.loadFile(VALID_CSV);
    controller.play();
    expect(states).toContain('Loaded');
    expect(states).toContain('Playing');
    for (const s of states) {
      expect(['Idle', 'Loaded', 'Playing', 'Paused', 'Error']).toContain(s);
    }
  });
});
